v2/comments.go

```diff
@@ -229,227 +225,185 @@ func (t Tag) String() string {
-	}
-	tagName := s.TokenText()
-	if len(tagNames) > 0 && !slices.Contains(tagNames, tagName) {
-		return "", nil, nil
-	}
-	if s.PeekSkipSpaces() != '(' {
-		return tagName, nil, nil
-	}
-	s.Scan() // consume the '(' token
-	args, err := s.parseTagArgs()
-	if err != nil {
-		return "", nil, err
-	}
-	if s.Scan() != ')' {
-		return "", nil, s.unexpectedTokenError("')'", s.TokenText())
-	}
-	return tagName, args, nil
-}
-
-// parseTagArgs parses the arguments part of an extended comment tag. The input
-// is assumed to be the entire text of the original input after the opening
-// '(', and before the trailing ')'.
-//
-// The argument may be a go style identifier, a quoted string ("..."), or a raw string (`...`).
-func (s *tagKeyScanner) parseTagArgs() ([]string, error) {
-	if s.PeekSkipSpaces() == ')' {
-		return nil, nil
-	}
-	if s.PeekSkipSpaces() == '{' || s.PeekSkipSpaces() == '[' {
-		value, err := s.scanJSONFlavoredValue()
-		if err != nil {
-			return nil, err
+//	Comment line without marker
+//	+k8s:required
+//	+listType=set
+//	+k8s:format=k8s-long-name
+//
+// Then this function will return:
+//
+//	map[TagIdentifier][]string{
+//		{Group: "k8s", Name: "required"}: {"required"},
+//		{Group: "k8s", Name: "required"}: {"format=k8s-long-name"},
+//	}
+func ExtractTags(marker string, group *string, lines []string) map[TagIdentifier][]string {
+	out := map[TagIdentifier][]string{}
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if !strings.HasPrefix(line, marker) {
+			continue
 		}
-		return []string{value}, nil
-	}
-	switch s.Scan() {
-	case scanner.String, scanner.RawString, scanner.Ident:
-		return []string{s.TokenText()}, nil
-	default:
-		return nil, s.unexpectedTokenError("identifier, quoted string (\"...\") or raw string (`...`)", s.TokenText())
-	}
-}
-
-// scanJSONFlavoredValue consumes a single token as a JSON value from the scanner and returns the token text.
-// A strict subset of JSON is supported, in particular:
-// - Big numbers and numbers with exponents are not supported.
-// - JSON is expected to be in a single line. Tabs and newlines are not fully supported.
-func (s *tagKeyScanner) scanJSONFlavoredValue() (string, error) {
-	start, end, err := s.chompJSONFlavoredValue()
-	if err != nil {
-		return "", err
-	}
-	value := s.input[start:end]
-	var out any
-	err = json.Unmarshal([]byte(value), &out) // make sure the JSON parses
-	if err != nil {
-		return "", err
-	}
-	return value, nil
-}
+		line = line[len(marker):]
 
-// chompJSONFlavoredValue consumes valid JSON from the scanner's token stream and returns the start and end positions of the JSON.
-func (s *tagKeyScanner) chompJSONFlavoredValue() (int, int, error) {
-	switch s.PeekSkipSpaces() {
-	case '[':
-		return s.chompJSONFlavoredArray()
-	case '{':
-		return s.chompJSONFlavoredObject()
-	}
+		nameEnd := len(line)
+		if idx := strings.IndexAny(line, "(="); idx > 0 {
+			nameEnd = idx
+		}
+		ident := TagIdentifierFromString(line[:nameEnd])
```

Comment:
If we keep "group", we could just move this to the state-machine -- would it be easier?

Reply:
I've dropped the group concept and allowed ":" in tag names.